File: serenityboard/writer/blob_storage.py
"""Content-addressed blob storage for SerenityBoard."""
from __future__ import annotations

import hashlib
import os
import sqlite3

__all__ = ["BlobStorage"]


class BlobStorage:
    """Content-addressed blob store backed by flat files.

    blob_key = sha256(data)[:16] + "." + extension
    Writes use tmp+rename for POSIX atomicity. Deduplicates on content hash.
    """

    def __init__(self, blobs_dir: str) -> None:
        """Create blob storage rooted at *blobs_dir* (created if missing)."""
        self._blobs_dir = blobs_dir
        os.makedirs(blobs_dir, exist_ok=True)

    def store(self, data: bytes, extension: str) -> str:
        """Store *data*, return blob_key. Skips write if already present."""
        blob_key = hashlib.sha256(data).hexdigest()[:16] + "." + extension
        path = os.path.join(self._blobs_dir, blob_key)
        if os.path.exists(path):
            return blob_key  # dedup
        tmp_path = path + ".tmp"
        with open(tmp_path, "wb") as f:
            f.write(data)
        os.rename(tmp_path, path)
        return blob_key

    def get_path(self, blob_key: str) -> str:
        """Return the full filesystem path for *blob_key*."""
        return os.path.join(self._blobs_dir, blob_key)

    def gc(self, conn: sqlite3.Connection) -> int:
        """Delete orphaned blob files not referenced in the DB.

        Returns the number of files removed.
        """
        live = {
            row[0]
            for row in conn.execute(
                "SELECT DISTINCT blob_key FROM artifacts"
            ).fetchall()
        }
        # Also include audio blobs so GC doesn't delete them
        try:
            live.update(
                row[0]
                for row in conn.execute(
                    "SELECT DISTINCT blob_key FROM audio"
                ).fetchall()
            )
        except sqlite3.OperationalError:
            pass  # audio table may not exist in older databases
        # Also include embedding blobs (matrix + sprite) so GC doesn't delete them
        try:
            live.update(
                row[0]
                for row in conn.execute(
                    "SELECT DISTINCT tensor_blob_key FROM embeddings"
                ).fetchall()
            )
            live.update(
                row[0]
                for row in conn.execute(
                    "SELECT DISTINCT sprite_blob_key FROM embeddings WHERE sprite_blob_key IS NOT NULL"
                ).fetchall()
            )
        except sqlite3.OperationalError:
            pass  # embeddings table may not exist in older databases
        # Also include mesh blobs (vertices, faces, colors) so GC doesn't delete them
        try:
            live.update(
                row[0]
                for row in conn.execute(
                    "SELECT DISTINCT vertices_blob_key FROM meshes"
                ).fetchall()
            )
            live.update(
                row[0]
                for row in conn.execute(
                    "SELECT DISTINCT faces_blob_key FROM meshes WHERE faces_blob_key IS NOT NULL"
                ).fetchall()
            )
            live.update(
                row[0]
                for row in conn.execute(
                    "SELECT DISTINCT colors_blob_key FROM meshes WHERE colors_blob_key IS NOT NULL"
                ).fetchall()
            )
        except sqlite3.OperationalError:
            pass  # meshes table may not exist in older databases
        # Also include graph blobs so GC doesn't delete them
        try:
            for row in conn.execute(
                "SELECT DISTINCT graph_blob_key FROM graphs"
            ).fetchall():
                # Skip legacy inline JSON values (not blob keys)
                if row[0] and not row[0].startswith("{"):
                    live.add(row[0])
        except sqlite3.OperationalError:
            pass  # graphs table may not exist or column may have old name
        removed = 0
        for entry in os.scandir(self._blobs_dir):
            if entry.is_file() and entry.name not in live:
                os.unlink(entry.path)
                removed += 1
        return removed

File: native/src/core/blob_storage.cpp
#include "serenityboard/blob_storage.hpp"

#include <cstdio>
#include <filesystem>
#include <fstream>
#include <stdexcept>

#include "serenityboard/hash.hpp"

namespace sb {

BlobStorage::BlobStorage(std::string blobs_dir) : blobs_dir_(std::move(blobs_dir)) {
  std::filesystem::create_directories(blobs_dir_);
}

std::string BlobStorage::key_for(std::string_view data, const std::string &extension) {
  const auto digest = sha256(data);
  return hex(digest.data(), digest.size()).substr(0, 16) + "." + extension;
}

std::string BlobStorage::store(std::string_view data, const std::string &extension) const {
  const std::string key = key_for(data, extension);
  const std::filesystem::path path = std::filesystem::path(blobs_dir_) / key;
  if (std::filesystem::exists(path)) return key;
  const std::filesystem::path tmp = path.string() + ".tmp";
  {
    std::ofstream out(tmp, std::ios::binary | std::ios::trunc);
    if (!out) throw std::runtime_error("cannot write blob " + tmp.string());
    out.write(data.data(), static_cast<std::streamsize>(data.size()));
    if (!out) throw std::runtime_error("short write for blob " + tmp.string());
  }
  std::error_code ec;
  std::filesystem::rename(tmp, path, ec);
  if (ec) throw std::runtime_error("cannot rename blob " + tmp.string() + ": " + ec.message());
  return key;
}

std::string BlobStorage::get_path(const std::string &key) const {
  return (std::filesystem::path(blobs_dir_) / key).string();
}

}  // namespace sb
